File: easy/SortByBits.cs
namespace Easy;

using System.Collections;
using System.Numerics;

public class SortByBits
{
    public int[] Solve(int[] arr)
    {
        var groups = new Dictionary<int, List<int>>();
        foreach (int num in arr)
        {
            int ones = BitOperations.PopCount((uint)num);
            if (!groups.TryGetValue(ones, out var list))
            {
                list = [];
                groups[ones] = list;
            }
            list.Add(num);
        }
        var result = new List<int>(arr.Length);
        foreach (var pair in groups.OrderBy(k => k.Key))
        {
            pair.Value.Sort();
            result.AddRange(pair.Value);
        }
        return [.. result];
    }
}

File: easy/sort_by_bits.go
package easy

import (
	"maps"
	"slices"
)

func CountSetBits(n int) int {
	count := 0
	for n != 0 {
		count += n & 1
		n >>= 1
	}
	return count
}
func SortByBits(arr []int) []int {
	freq := make(map[int][]int)
	for _, num := range arr {
		count := CountSetBits(num)
		freq[count] = append(freq[count], num)
	}
	for _, v := range freq {
		slices.Sort(v)
	}
	result := make([]int, 0, len(arr))
	for _, v := range slices.Sorted(maps.Keys(freq)) {
		result = append(result, freq[v]...)
	}
	return result
}
